### PR TITLE
Register the "dns.weave.local" name in WeaveDNS

### LINKED ISSUE #984: Register weaveDNS resolver address in weaveDNS
nginx can reverse proxy/load balance across servers by domain name as long as a resolver is specified explicitly in the configuration (system resolver cannot be used, presumably as it does not have a non-blocking API). At the moment this would require determining the IP address from `/etc/resolv.conf` and injecting it into the configuration, which is a little awkward. Fortunately the resolver address can itself be specified by hostname, and this is looked up (with the system resolver) when nginx starts. Consequently if each weaveDNS peer were to publish a well known name (e.g. `weavedns.weave.local`) pointing to the correct resolver IP, this name could be used in the nginx configuration and the system would bootstrap itself automatically with no need to preprocess configuration files.

### PR DESCRIPTION
**Pull request by inercia**
Fixes #984 by registering `dns.weave.local.` in a private ident that is not exported via mDNS, so containers can query about `dns.weave.local.` and get the IP of the local WeaveDNS instance.

### CODE REVIEW

File: test/200_dns_test.sh [hunk 1]
```diff
@@ -14,6 +14,7 @@ start_container          $HOST1 $C2/24 --name=c2 -h $NAME
 start_container_with_dns $HOST1 $C1/24 --name=c1
 
 assert_dns_record $HOST1 c1 $NAME $C2
+assert_dns_record $HOST1 c1 "dns.weave.local" 10.2.254.1
```
Comment (awh) [inline test/200_dns_test.sh]: Could we make this an variable to match the multicast tests

**Comment by awh** [comment]
Am ready to merge this once minor comment is addressed!

File: weave [hunk 2]
```diff
@@ -1027,6 +1027,8 @@ launch_dns() {
     ipam_cidrs_or_die $DNS_CONTAINER $CIDR_ARGS
     with_container_netns_or_die $DNS_CONTAINER attach $ALL_CIDRS
     populate_dns
+    # register "dns.weave.local." privately (not exported via mDNS)
+    put_dns_fqdn 'weave:private' dns.`dns_domain` $ALL_CIDRS
```
Comment (rade) [inline weave]: - single quotes aren't necessary here; I've [just](https://github.com/weaveworks/weave/commit/5559f1df2e4d50b4f08d600a633e57ced5490b6b) removed them elsewhere on master.
- never use backticks; always `$(...)`.

File: test/200_dns_test.sh [hunk 3]
```diff
@@ -5,15 +5,17 @@
...
 start_suite "Resolve names on a single host"
 
-launch_dns_on $HOST1 10.2.254.1/24
+launch_dns_on $HOST1 $DNS1/24
 
 start_container          $HOST1 $C2/24 --name=c2 -h $NAME
 start_container_with_dns $HOST1 $C1/24 --name=c1
 
 assert_dns_record $HOST1 c1 $NAME $C2
+assert_dns_record $HOST1 c1 "dns.weave.local" DNS1
```
Comment (awh) [inline test/200_dns_test.sh]: This test is failing - should be `$DNS1`

**Comment by squaremo** [comment]
Is there a reason to pollute the user-owned namespace `weave.local` with this name? Why not use something else, e.g., `.local` (I know, the user is allowed to change the name of the domain, but that does not excuse the pollution).

Also: unless WeaveDNS is in the same subnet as the container asking, they will not be able to talk. This seems like a bit of a sticking point for the issue which motivated this change.

**Comment by rade** [comment]
> Also: unless WeaveDNS is in the same subnet as the container asking, they will not be able to talk. This seems like a bit of a sticking point for the issue which motivated this change.

Why not populate it with the docker bridge ip? As per https://github.com/weaveworks/weave/issues/984#issuecomment-114425450

**Comment by squaremo** [comment]
> Why not populate it with the docker bridge ip? As per #984 (comment)

And if you're doing that, why not put an entry in `/etc/hosts` while you're at it (e.g., with `--add-host` or `--link`).

**Comment by rade** [comment]
@fintanr see above - could you give that a try?

**Comment by rade** [comment]
i.e. `docker run --add-host weavedns:$(weave docker-bridge-ip) ...` should do the trick and allow the nginx to resolve `weavedns`.

**Comment by fintanr** [comment]
@rade sure, starting a test/update for nginx guides now.

**Comment by squaremo** [comment]
> `docker run --add-host weavedns:$(weave docker-bridge-ip) ...`

or even

    docker run --link weavedns ...

HOWEVER: linking also copies the environment variables from the source container over to the target container (!?). So perhaps not recommended.

**Comment by inercia** [comment]
What is the problem with `polluting` the `weave.local.` domain? We are not introducing a random name. We are setting the dns server of that domain. It is a common convention to have a `dns.somedomain.com` server for the `somedomain.com` domain. Registering `dns` in a different domain provides little benefit and much complexity.

**Comment by squaremo** [comment]
> What is the problem with polluting the weave.local. domain?

It's meant to be under control of the user -- withholding a name in it is a bad smell.

Sure, it may be a convention for DNS admins to put dns.example.com -- but it's not done automatically by the nameserver (is it?). In any case I don't think we need this for fixing #984.

**Comment by inercia** [comment]
@squaremo So you are suggesting to fix this with `--add-host`. Taking into account that `nginx` uses its own resolver, have we verified that it reads `/etc/hosts`?

So I guess I should close this PR.

**Comment by squaremo** [comment]
> @squaremo So you are suggesting to fix this with --add-host. Taking into account that nginx uses its own resolver, have we verified that it reads /etc/hosts?

According to the description in #984, nginx uses the _system resolver_ first, to find the host you told it to use for _its own resolver_.

> So I guess I should close this PR.

Only if there's otherwise no use for this feature -- i.e., is there another scenario in which it would be handy to be able to resolve the DNS server itself?

**Comment by inercia** [comment]
> Only if there's otherwise no use for this feature -- i.e., is there another scenario in which it would be handy to be able to resolve the DNS server itself?

I think there isn't... I'm closing it.

**Comment by rade** [comment]
Let's wait for @fintanr to confirm that the `--add-host` approach actually works.

**Comment by inercia** [comment]
Any news on this, @fintanr?

**Comment by fintanr** [comment]
--add-host worked fine for me

**Comment by kpdecker** [comment]
For anyone who stumbles on this via google: IPv6 must be turned off for the resolver to work, at least as of Nginx 1.9.15.
https://trac.nginx.org/nginx/ticket/974

TL;DR: If the resolver setting is not working, try `ipv6=off` resolver flag.

**Comment by bboreham** [comment]
@kpdecker I believe you are describing the same problem as #2244, which is fixed now and will be in the next release.  Sorry for the trouble.

**Comment by kpdecker** [comment]
@bboreham thanks for the update. I have a work around for now and will certainly give the next version a try!

**Comment by bboreham** [comment]
@kpdecker we just released version 1.5.1 with the fix to #2244 included.

**Comment by kpdecker** [comment]
@bboreham thanks, will make some time to test it out.
